### PR TITLE
New package: CppAD

### PR DESCRIPTION
**Pull request by jppelteret**

### CODE REVIEW

File: var/spack/repos/builtin/packages/cppad/package.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    version('20170114.gpl', '565a534dc813fa1289764222cd8c11ea')
+    version('develop', git='https://github.com/coin-or/CppAD.git')
+
+    depends_on('cmake', type='build')
+
+    def cmake_args(self):
+        # This package does not obey CMAKE_INSTALL_PREFIX
+        args = [
+            "-Dcppad_prefix=%s" % (self.prefix),
+            "-Dcmake_install_docdir=%s" % ("share/cppad/doc")
```
Comment (davydden) [inline var/spack/repos/builtin/packages/cppad/package.py]: No need for prefix here, or is it relative to install dir?

Comment (jppelteret) [inline var/spack/repos/builtin/packages/cppad/package.py]: No, it turns out to be relative. I initially had a prefix here but the result wasn't as expected.

Comment (davydden) [inline var/spack/repos/builtin/packages/cppad/package.py]: I see, then u don't need '%s', just write the full string.

Comment (jppelteret) [inline var/spack/repos/builtin/packages/cppad/package.py]: Ah true. A legacy of my first attempt. Thanks!

File: var/spack/repos/builtin/packages/cppad/package.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+class Cppad(CMakePackage):
+    """A Package for Differentiation of C++ Algorithms."""
+
+    homepage = "https://www.coin-or.org/CppAD/"
+    url      = "http://www.coin-or.org/download/source/CppAD/cppad-20170114.gpl.tgz"
+
+    version('20170114.gpl', '565a534dc813fa1289764222cd8c11ea')
+    version('develop', git='https://github.com/coin-or/CppAD.git')
+
+    depends_on('cmake', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cppad/package.py]: This line is no longer required if you subclass `CMakePackage` as of #2623. I know `spack create` still adds it by default, but this is fixed in #2684.

File: var/spack/repos/builtin/packages/cppad/package.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+from spack import *
+
+
+class Cppad(CMakePackage):
+    """A Package for Differentiation of C++ Algorithms."""
+
+    homepage = "https://www.coin-or.org/CppAD/"
+    url      = "http://www.coin-or.org/download/source/CppAD/cppad-20170114.gpl.tgz"
+
+    version('20170114.gpl', '565a534dc813fa1289764222cd8c11ea')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cppad/package.py]: I see that there are both `gpl` and `cpl` downloads available. Out of curiosity, do you happen to know what the difference is? I'm wondering whether or not it is necessary to put that in the version number.

Comment (jppelteret) [inline var/spack/repos/builtin/packages/cppad/package.py]: Sorry, I don't know what the difference for the licence between the two versions is.  I tested without that part in the version string and found that it is unfortunately necessary (my commit is based off of `d3d87ea`). It appears to me that spack is not detecting that the file's full suffix is `.gpl.tgz`?

```bash
$ spack install cppad
==> Installing cppad
==> cmake is already installed in /Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-7.3.0-apple/cmake-3.6.1-dbk37r7pllmngnjzo4umf4hbszv2j7i7
==> Fetching http://www.coin-or.org/download/source/CppAD/cppad-20170114.tgz
######################################################################## 100.0%
curl: (22) The requested URL returned error: 404 Not Found
==> Fetching from http://www.coin-or.org/download/source/CppAD/cppad-20170114.tgz failed.
==> Error: FetchError: All fetchers failed for cppad-20170114-jm5ietsxi6c4zz53o3e3q4fhwe7udkvz
/Users/jp/Programs/GitHub/Spack/spack/lib/spack/spack/package.py:965, in do_fetch:
     938      def do_fetch(self, mirror_only=False):
     939          """
     940          Creates a stage directory and downloads the tarball for this package.
     941          Working directory will be set to the stage directory.
     942          """
     943          if not self.spec.concrete:
     944              raise ValueError("Can only fetch concrete packages.")
     945
     946          start_time = time.time()
     947          if spack.do_checksum and self.version not in self.versions:
     948              tty.warn("There is no checksum on file to fetch %s safely." %
     949                       self.spec.format('$_$@'))
     950
     951              # Ask the user whether to skip the checksum if we're
     952              # interactive, but just fail if non-interactive.
     953              ck_msg = "Add a checksum or use --no-checksum to skip this check."
     954              ignore_checksum = False
     955              if sys.stdout.isatty():
     956                  ignore_checksum = tty.get_yes_or_no("  Fetch anyway?",
     957                                                      default=False)
     958                  if ignore_checksum:
     959                      tty.msg("Fetching with no checksum.", ck_msg)
     960
     961              if not ignore_checksum:
     962                  raise FetchError("Will not fetch %s" %
     963                                   self.spec.format('$_$@'), ck_msg)
     964
  >> 965          self.stage.fetch(mirror_only)
     966
     967          self._fetch_time = time.time() - start_time
     968
     969          if spack.do_checksum and self.version in self.versions:
     970              self.stage.check()
     971
     972          self.stage.cache_local()
```

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cppad/package.py]: Yes, you would have to add a `url_for_version` to explain to Spack how to correctly handle the versioning. Not sure if is worth it just to make the version number simpler. We could rework how Spack parses versions for every little case like this, but there will always be a new scenario that we cannot predict.

**Comment by jppelteret** [comment]
Thanks a lot for the advice! I’d never used that functionality before. I’ve updated the PR with to use url_for_version.

**Pull request merged**